File: 1/2020.09.10/15.3Sum/python/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        l = len(nums)
        if l < 3:
            return []
        nums.sort()
        output = []
        for i in range(l-2):
            if nums[i] > 0:
                return output
            if i > 0 and nums[i] == nums[i-1]:
                continue
            target = -nums[i]
            ps, pe = i+1, l-1
            while ps < pe:
                total = nums[ps] + nums[pe]
                if total < target:
                    ps += 1
                elif total > target:
                    pe -= 1
                else:
                    output.append([-target, nums[ps], nums[pe]])
                    while ps < pe and nums[ps] == nums[ps+1]:
                        ps += 1
                    while pe > ps and nums[pe] == nums[pe-1]:
                        pe -= 1
                    ps += 1
                    pe -= 1
        return output

File: 1/2020.09.10/15.3Sum/java/solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        if (nums.length < 3) {
            return output;
        }
        Arrays.sort(nums);
        int l = nums.length;
        for (int i = 0; i < l - 2; i++) {
            if (nums[i] > 0) {
                return output;
            }
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            int target = -nums[i];
            int ps = i+1;
            int pe = l-1;
            while (ps < pe) {
                int total = nums[ps] + nums[pe];
                if (total < target) {
                    ps++;
                } else if (total > target) {
                    pe--;
                } else {
                    List<Integer> list = new ArrayList<>();
                    list.add(-target);
                    list.add(nums[ps]);
                    list.add(nums[pe]);
                    output.add(list);
                    while (ps < pe && nums[ps] == nums[ps+1]) {
                        ps++;
                    }
                    while (pe > ps && nums[pe] == nums[pe-1]) {
                        pe--;
                    }
                    ps++;
                    pe--;
                }
            }
        }
        return output;
    }
}